src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldServer.java

```diff
@@ -735,6 +738,34 @@ public void onUpdateTick(Block block, net.minecraft.world.World worldIn, BlockPo
 
     @Redirect(method = "addBlockEvent", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/WorldServer$ServerBlockEventList;add(Ljava/lang/Object;)Z", remap = false))
     public boolean onAddBlockEvent(WorldServer.ServerBlockEventList list, Object obj, BlockPos pos, Block blockIn, int eventId, int eventParam) {
+        // We fire a Pre event to make sure our captures do not get stuck in a loop.
+        // This is very common with pistons as they add block events while blocks are being notified.
+        if (blockIn instanceof BlockPistonBase) { // if it's a piston, try to simulate it
+            final IBlockState blockstate = this.getBlockState(pos);
+            EnumFacing direction = (EnumFacing) blockstate.getValue(BlockDirectional.FACING);
+
+            List<Location<org.spongepowered.api.world.World>> locations = new ArrayList<Location<org.spongepowered.api.world.World>>();
+            BlockPistonStructureHelper movedBlocks = new BlockPistonStructureHelper((WorldServer) (Object) this, pos, direction, eventId == 0);
+            if (movedBlocks.canMove()) { // calculates blocks to be moved
+                List<BlockPos> blocks = new ArrayList<BlockPos>();
+                blocks.addAll(movedBlocks.getBlocksToMove());
+                blocks.addAll(movedBlocks.getBlocksToDestroy());
```

Comment:
This needs to filter the block positions out, or provide a comment explaining that it already is filtered from the structure helper. We do not want to have duplicate positions in the event, especially since it's a simulated event, and not a final list of all blocks and snapshots changed.